### PR TITLE
adminserver: fix panic in DataDistribution

### LINKED ISSUE #30186: sql: panic: expected *DBytes, found *tree.DString
Occured on `cyan` running b6b6d0a
```
panic: expected *DBytes, found *tree.DString

goroutine 7963243 [running]:
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.MustBeDBytes(0x2fa4c80, 0xc4290bc850, 0xc4290bc850, 0x2fa4c80)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:1262 +0xe9
github.com/cockroachdb/cockroach/pkg/server.(*adminServer).DataDistribution(0xc42004fe40, 0x2fa22c0, 0xc42b9c0570, 0x4789ab0, 0xc42004fe40, 0xc425a74a08, 0xc42e660440)
        /go/src/github.com/cockroachdb/cockroach/pkg/server/admin.go:1542 +0xa8d
github.com/cockroachdb/cockroach/pkg/server/serverpb._Admin_DataDistribution_Handler.func1(0x2fa22c0, 0xc42b9c0570, 0x28a13a0, 0x4789ab0, 0x28a13a0, 0x4789ab0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/server/serverpb/admin.pb.go:1955 +0x86
github.com/cockroachdb/cockroach/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1(0x2fa22c0, 0xc42b9c0570, 0x28a13a0, 0x4789ab0, 0xc42b14d8a0, 0xc42b14d8c0, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc/server.go:48 +0xa33
github.com/cockroachdb/cockroach/pkg/rpc.NewServerWithInterceptor.func1(0x2fa22c0, 0xc42b9c0570, 0x28a13a0, 0x4789ab0, 0xc42b14d8a0, 0xc42b14d8c0, 0x26f2ce0, 0x4789ab0, 0x29f31e0, 0xc42a32e400)
        /go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:168 +0xb9
github.com/cockroachdb/cockroach/pkg/server/serverpb._Admin_DataDistribution_Handler(0x2a18c00, 0xc42004fe40, 0x2fa22c0, 0xc42b9c0570, 0xc42e9bf110, 0xc4203d1fc0, 0x0, 0x0, 0xc42a32e400, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/server/serverpb/admin.pb.go:1957 +0x167
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc420c16540, 0x2fc0ba0, 0xc4206f6000, 0xc42a32e400, 0xc420938f00, 0x41947e8, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1011 +0x4fc
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleStream(0xc420c16540, 0x2fc0ba0, 0xc4206f6000, 0xc42a32e400, 0x0)
        /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1249 +0x1318
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc4202ae060, 0xc420c16540, 0x2fc0ba0, 0xc4206f6000, 0xc42a32e400)
        /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:680 +0x9f
created by github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
        /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:678 +0xa1

```

### PR DESCRIPTION
**Pull request by rjnn**
WIP as I'm not 100% sure this is totally kosher, as this part of the codebase is new to me.

Also the fix is untested, I'd like to hear your opinion on how best to explicitly exercise this code in a test.

Closes #30186.

### CODE REVIEW

**Comment by arjunravinarayan** [comment]
@knz can you please review commit 1? Thank you.

**Comment by arjunravinarayan** [comment]
Thanks to @vilterp for helping me make this change!

**Comment by arjunravinarayan** [comment]
After talking to @jordanlewis, he's unhappy about the semantics change. So closing this in favor of #30201, which just hard changes the code to match the new schema.

**Review by knz — COMMENTED** [review]
So first of all I am guilty to have changed this interface so my apologies for giving you trouble.

Then this first commit doesn't work for me. The AsDBytes/MustBeDBytes (so far I can see) have also a type assertion contract, your change is dropping that assertion.

I would recommend making the type test (with a switch) in the admin endpoint directly. No need to change the sem/tree API.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30196#-:-LMIvJAX5RHDNUnZc8Xd:bkyfam8)* status: :shipit: complete! 0 of 0 LGTMs obtained

File: pkg/sql/sem/tree/datum.go [hunk 1]
```diff
@@ -1270,6 +1270,8 @@ func AsDBytes(e Expr) (DBytes, bool) {
 	switch t := e.(type) {
 	case *DBytes:
 		return *t, true
+	case *DString:
+		return DBytes(*t), true
```
Comment (jordanlewis) [inline pkg/sql/sem/tree/datum.go]: I don't think it's correct to do this - the contract of the `AsDFoo` methods seems to be to return a DFoo from an Expr if it is in fact a DFoo, and not to do any implicit casting. I don't think it's wise to add an implicit cast to just one of these.

**Pull request closed without merging**